hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/block/ProtoBlockFileReader.java

```
@@ -0,0 +1,253 @@
+                }
+
+                if (transaction != null) {
+                    var transactionResult = transactionResultProtoBlockItem.getTransactionResult();
+                    var blockItem = com.hedera.mirror.common.domain.transaction.BlockItem.builder()
+                            .transaction(transaction)
+                            .transactionResult(transactionResult)
+                            .transactionOutput(Collections.unmodifiableList(transactionOutputs))
+                            .stateChanges(Collections.unmodifiableList(stateChangesList))
+                            .build();
+                    context.getBlockFile()
+                            .addItem(blockItem)
+                            .onNewTransaction(getTransactionConsensusTimestamp(transactionResult));
+                }
+            } catch (InvalidProtocolBufferException e) {
+                throw new InvalidStreamFileException(
+                        "Failed to deserialize Transaction from block file " + context.getFilename(), e);
+            }
+        }
+    }
+
+    private void readRounds(ReaderContext context) {
+        BlockItem blockItem;
+        while (context.hasNextBlockItem() && (blockItem = context.readBlockItemFor(ROUND_HEADER)) != null) {
+            context.getBlockFile().onNewRound(blockItem.getRoundHeader().getRoundNumber());
+            readEvents(context);
+        }
+    }
+
+    /**
+     * Read trailing state changes. There is no marker to distinguish between transactional and non-transactional
+     * statechanges. This function reads those trailing non-transactional statechanges without immediately preceding
+     * transactional statechanges.
+     *
+     * @param context - The reader context
+     */
+    private void readTrailingStateChanges(ReaderContext context) {
+        while (context.hasNextBlockItem()) {
+            if (context.readBlockItemFor(STATE_CHANGES) == null) {
+                break;
+            }
+        }
```

Are the trailing state changes mandatory for gathering the output hashes and calculating the block hash? If so, maybe this method should throw an exception similar to the others if no trailing state changes are found.

---

no, not needed for us

---

We used to be very strict to fail on all incorrect data as much as we can detect in the stream files, however I believe we decided to move away from that approach. So that in the block file reader, unless the data is critical for mirrornode to properly parse and ingest the stream files, we should not fail / halt when missing / incorrect.